query-service-impl/src/main/java/org/hypertrace/core/query/service/promql/QueryRequestToPromqlConverter.java

```diff
@@ -0,0 +1,224 @@
+    this.viewDefinition = viewDefinition;
+  }
+
+  /**
+   * 1. selection should be equal to group by list
+   * 2. number of function selection is equal to number of query fired
+   * 3. count(*) type of query can't be served
+   */
+  PromqlQuery toPromql(
+      ExecutionContext executionContext,
+      QueryRequest request,
+      LinkedHashSet<Expression> allSelections) {
+    String groupByList = getGroupByList(request);
+
+    StringBuilder filter = new StringBuilder();
+    convertFilter2String(request.getFilter(), filter);
+
+
+    QueryTimeRange queryTimeRange = executionContext.getQueryTimeRange().orElseThrow(
+        () -> new RuntimeException("Time Range missing in query"));
+
+    List<String> queries = allSelections.stream().filter(expression -> expression.getValueCase().equals(ValueCase.FUNCTION))
+        .map(functionExpression -> {
+          String functionName = getFunctionName(functionExpression);
+          String metricName = getMetricNameFromFunction(functionExpression);
+          return buildQuery(metricName, functionName, groupByList, filter.toString(), queryTimeRange.getDuration().toMillis());
+        }).collect(Collectors.toUnmodifiableList());
+
+    return new PromqlQuery(
+        queries, executionContext.getTimeSeriesPeriod().isEmpty(),
+        queryTimeRange.getStartTime(), queryTimeRange.getEndTime(),
+        getStep(executionContext));
+  }
+
+  private String getGroupByList(QueryRequest queryRequest) {
+    List<String> groupByList = Lists.newArrayList();
+    if (queryRequest.getGroupByCount() > 0) {
+      for (Expression expression : queryRequest.getGroupByList()) {
+        groupByList.add(convertColumnIdentifierExpression2String(expression));
+      }
+    }
+    return String.join(",", groupByList);
+  }
+
+  private long getStep(ExecutionContext executionContext) {
+    return executionContext
+        .getTimeSeriesPeriod().map(Duration::toMillis)
+        .orElse(-1L);
+  }
+
+  private String buildQuery(
+      String metricName, String function,
+      String groupByList, String filter, long durationMillis) {
+    String template = "%s by (%s) (%s(%s{%s}[%sms]))"; // sum by (a1, a2) (sum_over_time(num_calls{a4="..", a5=".."}[xms]))
+
+    return String.format(template, function, groupByList, function + "_over_time", metricName, filter, durationMillis);
+  }
+
+  private String getFunctionName(
+      Expression functionSelection) {
+    switch (functionSelection.getFunction().getFunctionName().toUpperCase()) {
+      case QUERY_FUNCTION_SUM:
+        return "sum";
+      case QUERY_FUNCTION_MAX:
+        return "max";
+      case QUERY_FUNCTION_MIN:
+        return "min";
+      case QUERY_FUNCTION_AVG:
+        return "avg";
+      default:
+        throw new RuntimeException("");
+    }
+  }
+
+  private String getMetricNameFromFunction(
+      Expression functionSelection) {
+    // todo map columnName
+    return functionSelection.getColumnIdentifier().getColumnName();
```

Comment:
We are in a process of eventually deprecating `ColumnIdentifier` . So, let's add support for `AttributeExpression` as well.

Reply:
Added